File: Server/Context/src/main/java/org/elkoserver/server/context/CheckpointTask.java
package org.elkoserver.server.context;

abstract class CheckpointTask {
    private CheckpointTask() {
    }

    abstract void execute(Contextor contextor);

    static CheckpointTask makeDeleteTask(String ref) {
        return new DeleteTask(ref);
    }

    static CheckpointTask makeWriteTask(String ref, BasicObject obj) {
        return new WriteTask(ref, obj);
    }

    private static class DeleteTask extends CheckpointTask {
        private String myRef;
        DeleteTask(String ref) {
            myRef = ref;
        }
        void execute(Contextor contextor) {
            contextor.writeObjectDelete(myRef);
        }
    }

    private static class WriteTask extends CheckpointTask {
        private String myRef;
        private BasicObject myObj;
        WriteTask(String ref, BasicObject obj) {
            myRef = ref;
            myObj = obj;
        }
        void execute(Contextor contextor) {
            contextor.writeObjectState(myRef, myObj);
        }
    }
}

File: Server/Context/src/main/kotlin/org/elkoserver/server/context/CheckpointTask.kt
package org.elkoserver.server.context

internal abstract class CheckpointTask private constructor() {
    abstract fun execute(contextor: Contextor)
    internal class DeleteTask(private val myRef: String) : CheckpointTask() {
        override fun execute(contextor: Contextor) {
            contextor.writeObjectDelete(myRef)
        }
    }

    internal class WriteTask(private val myRef: String, private val myObj: BasicObject) : CheckpointTask() {
        override fun execute(contextor: Contextor) {
            contextor.writeObjectState(myRef, myObj)
        }
    }
}
